cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherClientHandler.java

```diff
@@ -0,0 +1,119 @@
+  private final CConfiguration cConf;
+  private final String instanceName;
+
+  @Inject
+  TetherClientHandler(CConfiguration cConf, TetherStore store) {
+    this.store = store;
+    this.cConf = cConf;
+    this.instanceName = cConf.get(Constants.INSTANCE_NAME);
+  }
+
+  @Override
+  public void init(HandlerContext context) {
+    super.init(context);
+    Class<? extends RemoteAuthenticator> authClass = cConf.getClass(Constants.Tether.CLIENT_AUTHENTICATOR_CLASS,
+                                                                    null,
+                                                                    RemoteAuthenticator.class);
+    if (authClass != null) {
+      try {
+        RemoteAuthenticator.setDefaultAuthenticator(authClass.newInstance());
+      } catch (Exception e) {
+        LOG.error("Failed to set default authenticator", e);
+      }
+    }
+  }
+
+  /**
+   * Initiates tethering with the server.
+   */
+  @POST
+  @Path("/tethering/create")
+  public void createTether(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetherCreationRequest tetherCreationRequest = GSON.fromJson(content, TetherCreationRequest.class);
+
+    List<PeerInfo> peers = store.getPeers();
+    Optional<PeerInfo> peer = peers.stream()
+      .filter(p -> p.getName().equals(tetherCreationRequest.getPeer()))
+      .findFirst();
+    if (peer.isPresent()) {
+      LOG.info("Peer {} is already present in state {}, ignoring tethering request",
+               peer.get().getName(), peer.get().getTetherStatus());
+      responder.sendStatus(HttpResponseStatus.OK);
+      return;
+    }
+
+    List<NamespaceAllocation> namespaces = tetherCreationRequest.getNamespaceAllocations();
+    TetherConnectionRequest tetherConnectionRequest = new TetherConnectionRequest(instanceName,
+                                                                                  namespaces);
+    if (tetherCreationRequest.getEndpoint() == null) {
+      throw new BadRequestException("Endpoint is null");
+    }
+    URI endpoint = new URI(tetherCreationRequest.getEndpoint());
+    HttpResponse response = TetherUtils.sendHttpRequest(HttpMethod.POST, endpoint.resolve(CREATE_TETHER),
+                                                        GSON.toJson(tetherConnectionRequest));
+    if (response.getResponseCode() != 200) {
+      LOG.error("Failed to send tether request, body: {}, code: {}",
+                response.getResponseBody(), response.getResponseCode());
+      responder.sendStatus(HttpResponseStatus.valueOf(response.getResponseCode()));
+      return;
+    }
+
+    PeerMetadata peerMetadata = new PeerMetadata(namespaces, tetherCreationRequest.getMetadata());
+    PeerInfo peerInfo = new PeerInfo(tetherCreationRequest.getPeer(), tetherCreationRequest.getEndpoint(),
+                                     TetherStatus.PENDING, peerMetadata);
+    store.addPeer(peerInfo);
```

Comment:
So what happen if we successfully sent the request to the server side, but failed to persist it locally? In the RemoteAgentService, I don't see this case is being handled there.

Reply:
Handled this case. Client side will need to retry tethering. Duplicate tethering request is handled by server side.